base/multidimensional.jl

```diff
@@ -690,14 +687,15 @@ julia> diff(vec(a))
  12
 ```
 """
-function diff(A::AbstractMatrix; dims::Integer)
-    if dims == 1
-        [A[i+1,j] - A[i,j] for i=1:size(A,1)-1, j=1:size(A,2)]
-    elseif dims == 2
-        [A[i,j+1] - A[i,j] for i=1:size(A,1), j=1:size(A,2)-1]
-    else
-        throw(ArgumentError("dimension must be 1 or 2, got $dims"))
-    end
+function diff(a::AbstractArray; dims::Integer)
+  @assert !has_offset_axes(a)
+  1 <= dims <= ndims(a) || throw(ArgumentError("dimension out of range"))
```

Comment:
Since you're using `ndims(a)` multiple times, it might be a little nicer to do
```julia
function diff(a::AbstractArray{T,N}; dims::Integer) where {T,N}
    ...
    1 <= dims <= N || ...
    ...
end
```

Reply:
BTW, printing the dimension which was actually passed (as the existing code does) is also nice for users.

Reply:
good call, fixed